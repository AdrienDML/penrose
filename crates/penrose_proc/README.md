## proc-macros for the Penrose tiling window manager library

Most (if not all) of the proc-macros in this crate are consumed internally in
the main [penrose][1] crate and are not intended for use outside of that
context. Please see the main penrose crate for further details.

  [1]: https://crates.io/crates/penrose
