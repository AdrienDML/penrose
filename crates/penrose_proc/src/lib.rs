//! Proc macros for use in the main Penrose crate
use proc_macro::TokenStream;

mod stub;
mod validate_bindings;

use stub::stubbed_companion_trait_inner;
use validate_bindings::validate_user_bindings_inner;

/// This is an internal macro that is used as part of `gen_keybindings` to validate user provided
/// key bindings at compile time using xmodmap.
///
/// It is not intended for use outside of that context and may be modified and updated without
/// announcing breaking API changes.
///
/// ```no_run
/// # use penrose_proc::validate_user_bindings;
/// validate_user_bindings!(
///     ( "M-a", "M-b" )
///     (
///         ( ( "M-{}", "M-S-{}" ) ( "1", "2", "3" ) )
///     )
/// );
/// ```
#[proc_macro]
pub fn validate_user_bindings(input: TokenStream) -> TokenStream {
    validate_user_bindings_inner(input)
}

#[proc_macro_attribute]
pub fn stubbed_companion_trait(args: TokenStream, input: TokenStream) -> TokenStream {
    stubbed_companion_trait_inner(args, input)
}
