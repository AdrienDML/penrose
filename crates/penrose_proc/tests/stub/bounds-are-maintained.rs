// The stub trait is generated with the same supertraits
use penrose_proc::stubbed_companion_trait;

#[stubbed_companion_trait]
trait Foo: std::fmt::Debug + Eq {
    #[stub("red")]
    fn color(&self, x: u32) -> &str;
}

#[derive(Debug)]
struct MyStruct;

impl StubFoo for MyStruct {}

fn main() {
    let s = MyStruct;
    assert_eq!(s.color(42), "red");
}
