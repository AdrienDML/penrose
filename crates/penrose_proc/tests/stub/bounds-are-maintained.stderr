error[E0277]: the trait bound `MyStruct: Eq` is not satisfied
  --> $DIR/bounds-are-maintained.rs:13:6
   |
5  | trait Foo: std::fmt::Debug + Eq {
   |                              -- required by this bound in `StubFoo`
...
13 | impl StubFoo for MyStruct {}
   |      ^^^^^^^ the trait `Eq` is not implemented for `MyStruct`
